src/main/kotlin/com/exactpro/th2/common/utils/message/MessageBatcher.kt

```diff
@@ -18,36 +18,36 @@ val RAW_DIRECTION_SELECTOR: (RawMessage.Builder) -> Any = { it.sessionAlias to i
 val DIRECTION_SELECTOR: (Message.Builder) -> Any = { it.sessionAlias to it.direction }
 
 class RawMessageBatcher(
-    maxBatchSize: Int = 1000,
     maxFlushTime: Long = 1000,
     private val batchSelector: (RawMessage.Builder) -> Any,
+    addGroup: MessageGroupBatch.Builder.(MessageGroup) -> Boolean,
     executor: ScheduledExecutorService,
     onError: (Throwable) -> Unit = {},
     onBatch: (MessageGroupBatch) -> Unit
-): Batcher<RawMessage.Builder>(maxBatchSize,maxFlushTime, executor, onError, onBatch) {
+): Batcher<RawMessage.Builder>(maxFlushTime, addGroup, executor, onError, onBatch) {
     override fun onMessage(message: RawMessage.Builder) {
         message.metadataBuilder.timestamp = Instant.now().toTimestamp()
         add(batchSelector(message), message.build())
     }
 }
 
 class MessageBatcher(
-    maxBatchSize: Int = 1000,
     maxFlushTime: Long = 1000,
     private val batchSelector: (Message.Builder) -> Any,
+    addGroup: MessageGroupBatch.Builder.(MessageGroup) -> Boolean,
     executor: ScheduledExecutorService,
     onError: (Throwable) -> Unit = {},
     onBatch: (MessageGroupBatch) -> Unit
-): Batcher<Message.Builder>(maxBatchSize,maxFlushTime, executor, onError, onBatch) {
+): Batcher<Message.Builder>(maxFlushTime, addGroup, executor, onError, onBatch) {
     override fun onMessage(message: Message.Builder) {
         message.metadataBuilder.timestamp = Instant.now().toTimestamp()
         add(batchSelector(message), message.build())
     }
 }
 
 abstract class Batcher<T>(
-    private val maxBatchSize: Int = 1000,
     private val maxFlushTime: Long = 1000,
+    private val addGroup: MessageGroupBatch.Builder.(MessageGroup) -> Boolean,
```

Comment:
I've did a little digging and looks like we can avoid passing whole builder to a user-function and can pass groups-list from it instead (it seems to be unmodifiable). This way user-function won't have ability to mutate batch and still can check if it can fit a new group in it